src/shared/ui/pickerPrompter.ts

```diff
@@ -0,0 +1,255 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+
+import { WIZARD_BACK } from '../wizards/wizard'
+import { QuickInputButton, PrompterButtons } from './buttons'
+import { Prompter, PromptResult } from './prompter'
+import { applyPrimitives } from '../utilities/collectionUtils'
+
+/** Additional options to configure the `QuickPick` beyond the standard API */
+interface AdditionalQuickPickOptions<T = never> {
+    title?: string
+    value?: string
+    step?: number
+    placeholder?: string
+    totalSteps?: number
+    buttons?: PrompterButtons<T>
+    customInputSettings?: {
+        label: string
+        transform: (v: string) => PromptResult<T>
+    }
+}
+
+export type ExtendedQuickPickOptions<T> = Omit<vscode.QuickPickOptions, 'buttons' | 'canPickMany' | 'placeHolder'> &
+    AdditionalQuickPickOptions<T>
+
+export const DEFAULT_QUICKPICK_OPTIONS: vscode.QuickPickOptions = {
+    ignoreFocusOut: true,
+}
+
+type QuickPickData<T> = PromptResult<T> | (() => Promise<PromptResult<T>>)
+type LabelQuickPickItem<T> = vscode.QuickPickItem & { label: T }
+
+/**
+ * Attaches additional information as `data` to a QuickPickItem. Alternatively, `data` can be a function that
+ * returns a Promise, evaluated after the user selects the item.
+ */
+export type DataQuickPickItem<T> = vscode.QuickPickItem & { data: QuickPickData<T> }
+export type DataQuickPick<T> = Omit<vscode.QuickPick<DataQuickPickItem<T>>, 'buttons'> & { buttons: PrompterButtons<T> }
+
+const CUSTOM_USER_INPUT = Symbol()
+
+function isDataQuickPickItem(obj: any): obj is DataQuickPickItem<any> {
+    return typeof obj === 'object' && typeof (obj as vscode.QuickPickItem).label === 'string' && 'data' in obj
+}
+
+/**
+ * Creates a new QuickPick using special DataQuickPickItem interfaces. Information that should be returned when
+ * the user selects an item should be placed in the `data` property of each item. If only the `label` is desired,
+ * use `createLabelQuickPick` instead.
+ *
+ * @param items An array or a Promise for items.
+ * @param options Customizes the QuickPick and QuickPickPrompter.
+ * @returns A QuickPickPrompter. This can be used directly with the `prompt` method or can be fed into a Wizard.
+ */
+export function createQuickPick<T>(
```

Comment:
Since we have named our extended QuickPick concept "Prompter" (and this returns `QuickPickPrompter`), does it make sense to name this`createPrompter` ? Same for `createLabelQuickPick`, etc.

Reply:
I've struggled with figuring out a good name for many of the functions/classes in this file. While yes the two factory functions return an instance of `Prompter`, they don't necessarily return the same kind of `Prompter`. I used `createQuickPick` because it's still a QuickPick that's displayed to the end-user, just internally there's layers of abstractions over the QuickPick object itself.

Reply:
Yeah that makes sense. I think the current name is fine. I think the main problem is that we are struggling with the layers of abstraction. `Prompter` is very abstract, it is basically a callback, but its name implies UI coupling.

Reply:
Agreed. `Prompter` is very abstract and contains no UI implementation details. The only references to UI are its public abstract methods. The fact that I've included many transformation callbacks within the abstraction also muddies its purpose. I think the ideal case would be to separate the transformation logic out and leave `Prompter` purely as an interface. Then there can be an abstract base `Transformable` class containing the transformation logic.